Sprint 6

baby-benz opened the pull request:
1. Добавлена поддержка прокси.
2. Исправлен таймер количества распаршенных сущностей в секунду.
3. Переписан cтатус-сервис для небольшого увеличения производительности.
4. Теперь шедулер находится в отдельном компоненте и вместо запуска в отдельном потоке в случае, если процесс парсинга уже идет, ожидает окончания парсинга.
5. Исправлена передача CSV-файла и ее эндпоинт.
6. Добавлена поддержка многопоточности. Взятие HTML-документа и парсинг отдельного продукта теперь происходят в самостоятельных потоках.
7. Логи об отсутсвии некоторых свойств теперь сопровождаются ссылками на этот продукт, т.к. логи разных продуктов идут вперемешку из-за внедрения многопоточности.
8. Добавлен пул объектов сервиса забора HTML-документов, т.к. разные потоки теперь конфликтуют за этот ресурс, а блокировка этого ресурса отрицательно сказывается на производительности. Вместо этого каждому потоку, которому необходим забор документов, предоставляется свой собственный экземпляр сервиса.
9. И другие множественные фиксы.

kav128 committed:
add proxy rotation

kav128 committed:
add proxy rotation

kav128 committed:
add web page loader abstraction

kav128 committed:
add implementation

kav128 committed:
return SocketTimeoutException catching

kav128 committed:
clean the code & remove rebase errors

kav128 committed:
Merge remote-tracking branch 'microservices-course-itmo/proxies' into proxies

# Conflicts:
#	src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ProxyService.java
#	src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ScrapingService.java

baby-benz committed:
force flyway to create sequence required for identity generation

baby-benz committed:
simplify id generation via SERIAL PostgreSQL datatype and Hibernate's GenerationType.IDENTITY

baby-benz committed:
add image table creation to flyway

Commit:
Integration Tests added

kanootoko committed:
Delete settings.json

This is from VS Code configuration, deleting

kanootoko committed:
Update application.properties

Restored application.properties

kanootoko committed:
Merge pull request #5 from microservices-course-itmo/integ_test

Integration Tests added

kanootoko committed:
Add some statictics about errors while parsing

kanootoko committed:
Change version to 0.6.0 for the 6th sprint

kav128 committed:
update ProxyService

kav128 committed:
fix codestyle

kav128 committed:
change static ScrapingService members to non-static

kav128 committed:
log exceptions

kav128 committed:
use Collections.emptyList()

kav128 committed:
add timeout

kav128 committed:
use timeout in parser

kav128 committed:
Merge pull request #6 from microservices-course-itmo/proxies

add proxy

baby-benz committed:
fix parsing timer seconds part

baby-benz committed:
fix parsing timer seconds part

baby-benz committed:
refactor StatusService for better performance

baby-benz committed:
refactor StatusService for better performance

baby-benz committed:
separate scheduler from ParsingControllerService and upgrade it
replace Thread manual executing with ExecutorService

baby-benz committed:
separate scheduler from ParsingControllerService and upgrade it
replace Thread manual executing with ExecutorService

baby-benz committed:
revert commits referring to scheduler upgrade due to wrong branch

baby-benz committed:
revert commits referring to scheduler upgrade due to wrong branch

kanootoko committed:
Fix CSV output and logging
Fix CSV to output UTF-8 correctly
Fix parsig logs to output URL of entity with missing fields

baby-benz committed:
add multithreading support to ParserService

baby-benz committed:
Merge branch 'sprint-6' into multithreading

# Conflicts:
#	src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/parser/ParserService.java

baby-benz committed:
fix exception catching in ScrapingService

baby-benz committed:
replace newSingleThreadExecutor with simple new thread starting in ParsingControllerService
fix WinestyleParserService interface not throwing InterruptedException

baby-benz committed:
remove redundant logging from ParsingController

baby-benz committed:
fix parsed wine logging

baby-benz committed:
import fix

baby-benz committed:
change naming

baby-benz committed:
upgrade new ExecutorService creating/shutdown

baby-benz committed:
fix job class naming

baby-benz committed:
add minor ProductJob optimizations

baby-benz committed:
Merge pull request #10 from microservices-course-itmo/multithreading

Multithreading

baby-benz committed:
add flyway migration to truncate database for performance testing

baby-benz committed:
pass pre-parsed url to logger instead of method calling

baby-benz committed:
Merge pull request #9 from microservices-course-itmo/sprint-6csv-fix

Fix CSV output and logging

baby-benz committed:
fix threads ScrapingService resource race problem

baby-benz committed:
fix pom duplicates

baby-benz committed:
move methods only used in ProductJob to ProductJob

kav128 committed:
fix WinestyleParserServiceMetricsCollector

kav128 committed:
use underscore in SERVICE_NAME

baby-benz committed:
Merge pull request #12 from microservices-course-itmo/s6-hotfix

build hotfix

kav128 committed:
add parameter for max timeout

AfanasievYuriy commented on pom.xml:
```diff
@@ -90,19 +86,38 @@
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-test</artifactId>
+			<scope>test</scope>
+		</dependency>
```
форматирование съехало

AfanasievYuriy commented on pom.xml:
```diff
@@ -90,19 +86,38 @@
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>2.8.6</version>
```
версию в проперти

AfanasievYuriy commented on src/main/java/com/wine/to/up/winestyle/parser/service/controller/MainController.java:
```diff
@@ -80,15 +81,18 @@ public Alcohol getAlcoholById(@PathVariable long id) throws NoEntityException {
                 try {
                     res.put(fieldName, field.get(alcohol));
-                } catch (IllegalArgumentException | IllegalAccessException ignore) { }
+                } catch (IllegalArgumentException | IllegalAccessException ignore) {
```
ребят, если уж catch блок есть, то добавьте туда какое-то поведение (минимум - лог, а желательно какой-то осознанный ответ клиенту)

kanootoko replied:
Блок есть, но он в принципе не может сработать, поэтому из него и убрано любое поведение. Мы бежим циклом по полям, делаем их доступными, и обращаемся. IllegalArgumentException невозможен, так как поле взято из списка полей, а IllegalAccessException невозможен, раз мы дали доступ

AfanasievYuriy commented on pom.xml:
```diff
@@ -15,14 +15,15 @@
+        <current-version>0.6.0</current-version>
 
-        <service.commonlib.version>0.2.1-SNAPSHOT</service.commonlib.version>
+        <service.commonlib.version>0.2.2-SNAPSHOT</service.commonlib.version>
```
обновитесь до последней версии

kav128 replied:
0.2.2-SNAPSHOT - и есть последняя версия для service-common-lib

AfanasievYuriy commented on pom.xml:
```diff
@@ -15,14 +15,15 @@
 
-        <service.commonlib.version>0.2.1-SNAPSHOT</service.commonlib.version>
+        <service.commonlib.version>0.2.2-SNAPSHOT</service.commonlib.version>
         <parser.common.api.version>0.2.1-SNAPSHOT</parser.common.api.version>
```
обновитесь до последней версии

baby-benz replied:
Но ведь это актуальная версия

baby-benz committed:
fix pom formatting

baby-benz committed:
add new IllegalFieldException
replace exception ignorance with logging and client message throwing
move client request handling logic into a separate class

kav128 committed:
update parser-common-api version

baby-benz committed:
add some refactoring
add minor optimizations
add additional information to finished parsing logging

baby-benz committed:
fix missed rating property logging

baby-benz committed:
Merge remote-tracking branch 'origin/sprint-6-hotfix' into sprint-6-hotfix

baby-benz committed:
Merge pull request #13 from microservices-course-itmo/sprint-6-hotfix

Sprint 6 hotfix

baby-benz committed:
update kafka configuration

baby-benz committed:
Merge remote-tracking branch 'origin/sprint-6' into sprint-6

baby-benz committed:
fix kafka image passing

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/controller/exception/GlobalExceptionHandler.java:
```diff
@@ -46,14 +47,23 @@
             return handleUnsupportedAlcoholTypeException(unsupportedAlcoholTypeException, headers, status, request);
         }
+
+        if (ex instanceof IllegalFieldException) {
```
Так не стоит делать, просто отдельный метод @ExceptionHandler(IllegalFieldException.class)

baby-benz committed:
fix IllegalFieldException throwing

baby-benz committed:
fix image kafka passing

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ProxyService.java:
```diff
@@ -0,0 +1,88 @@
+
+    private List<String> getAllProxies() {
+        try {
+            URL url = new URL("https://api.proxyscrape.com/?request=getproxies&proxytype=socks5&country=all");
```
Надо куда-то это вынести. Наример, в проперти

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ProxyService.java:
```diff
@@ -0,0 +1,88 @@
+        try {
+            URL url = new URL("https://api.proxyscrape.com/?request=getproxies&proxytype=socks5&country=all");
+            URLConnection connection = url.openConnection();
+            BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()));
```
try with resources

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ProxyService.java:
```diff
@@ -0,0 +1,88 @@
+        Proxy proxy = convertProxy(proxyAddress);
+
+        try {
+            Jsoup.connect("https://winestyle.ru/").proxy(proxy).timeout(maxTimeout).get();
```
Ссылку на сайт тоже лучше не так, вы же не в одном месте это используете

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ProxyService.java:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    private IUnstableLoader getProxyIfAlive(String proxyAddress, int maxTimeout) {
```
Мне не очень нравится) Может, getAliveProxy было бы лучше?

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ProxyService.java:
```diff
@@ -0,0 +1,88 @@
+            log.trace("{} OK", proxyAddress);
+            return new ProxyWebPageLoader(proxy);
+        } catch (Exception e) {
+            return null;
```
Null не очень хорошо возвращать. Можно попробовать optional

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ProxyService.java:
```diff
@@ -0,0 +1,88 @@
+        List<String> proxyAddresses = getAllProxies();
+        ExecutorService threadPool = Executors.newFixedThreadPool(proxyAddresses.size());
+        log.info("Loaded list of {} proxies. Checking", proxyAddresses.size());
+        futures = proxyAddresses.stream().map(proxyAddress -> CompletableFuture.supplyAsync(() -> getProxyIfAlive(proxyAddress, maxTimeout), threadPool)).collect(Collectors.toList());
```
Можно отдельно не объявлять)

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ScrapingService.java:
```diff
@@ -26,21 +36,26 @@ public Document getJsoupDocument(String url) throws InterruptedException {
+            catch (HttpStatusException e) {
                 log.error("An error occurs whilst fetching the URL!", e);
+            } catch (IOException e) {
+                e.printStackTrace();
```
Логгер)

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/ScrapingService.java:
```diff
@@ -26,21 +36,26 @@ public Document getJsoupDocument(String url) throws InterruptedException {
             }
         }
-        Thread.sleep(625);
+        if (timeout > 0) Thread.sleep(timeout);
```
Нехорошо в одной строчке

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/document/SimpleWebPageLoader.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public String toString() {
+        return "SimpleWebPageLoader{}";
```
Это зачем?)

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/parser/ParserService.java:
```diff
@@ -12,114 +15,226 @@
+        parsingThreadPool = Executors.newFixedThreadPool(MAX_THREAD_COUNT, parsingThreadFactory);
+    }
+
+    @SneakyThrows
```
Нехорошо, не используйте лучше

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/parser/ParserService.java:
```diff
@@ -12,114 +15,226 @@
+        }
+    }
+
+    @SneakyThrows
```
И здесь

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/parser/ParserService.java:
```diff
@@ -12,114 +15,226 @@
+    }
 
-        log.warn("Finished parsing of {} in {}", alcoholType, java.time.Duration.between((start), LocalDateTime.now()));
+    @SneakyThrows
```
зд

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/scheduling/AlcoholParsingScheduler.java:
```diff
@@ -0,0 +1,37 @@
+public class AlcoholParsingScheduler {
+    private final ParsingControllerService parsingControllerService;
+
+    @SneakyThrows(UnsupportedAlcoholTypeException.class)
```
з

shinehlina commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/scheduling/AlcoholParsingScheduler.java:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    @SneakyThrows(UnsupportedAlcoholTypeException.class)
```
з

shinehlina commented on src/test/java/com/wine/to/up/winestyle/parser/service/TestingWebApplicationTest.java:
```diff
@@ -0,0 +1,113 @@
+import com.wine.to.up.winestyle.parser.service.service.implementation.repository.AlcoholRepositoryService;
+
+@SpringBootTest
+@ComponentScan("com.wine.to.up")
```
Без этого не работает?

baby-benz replied:
Нет, без этого не подтянет ParserCommonApiProperties и ParserApi, потому что они находятся в com.wine.to.up.parser, а проект в com.wine.to.up.winestyle

shinehlina commented on src/test/java/com/wine/to/up/winestyle/parser/service/TestingWebApplicationTest.java:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	// @Test
+	// public void alcoholByURLSuccess() throws Exception {
```
Закомментированный?

shinehlina commented on src/test/resources/application.properties:
```diff
@@ -0,0 +1,18 @@
+management.endpoint.prometheus.enabled=true
+
+spring.datasource.url=jdbc:postgresql://${S_POSTGRES_HOST:localhost}:5432/postgres
+spring.datasource.username=postgres
```
Вы собираетесь тестировать на реальной базе?

shinehlina commented on src/test/resources/application.properties:
```diff
@@ -0,0 +1,18 @@
+spring.flyway.baseline-on-migrate=true
+
+# scheduler's properties for activate (00:00:00 everyday)
+scheduler.cron.expression=0 0 0 * * *
```
И со скедулингом?

shinehlina commented on src/test/resources/application.properties:
```diff
@@ -0,0 +1,18 @@
+spring.kafka.bootstrap-server=${S_KAFKA_BOOTSTRAP_HOST:${default.kafka.bootstrap-server}}
+
+# flyway properties - https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-application-properties.html#data-migration-properties
+spring.flyway.baseline-on-migrate=true
```
И с миграцией?

shinehlina commented on src/test/resources/application.properties:
```diff
@@ -0,0 +1,18 @@
+spring.application.name=winestyle-parser-service
+spring.kafka.consumer.group-id=wine.winestyle-parser-service
+
+management.endpoints.web.exposure.include=prometheus
```
И с промитиусом? Вставьте только нужные

shinehlina reviewed (changes requested):
Ой, кажется, я не засабмитила

baby-benz committed:
add lombok annotation in place of constructor

baby-benz committed:
add lombok annotation in place of constructor

baby-benz committed:
refactor GlobalExceptionHandler handlers

baby-benz committed:
Merge pull request #14 from microservices-course-itmo/sprint-6-hotfix

Sprint 6 hotfix

kav128 committed:
move proxy list url to properties

kav128 committed:
add proxy init call

baby-benz committed:
turn alcohol's strength into float

baby-benz committed:
bugfix strength migration files

baby-benz committed:
remove unsafe using of SneakyThrows annotation

baby-benz committed:
remove unsafe using of SneakyThrows annotation

baby-benz committed:
remove unsafe using of SneakyThrows annotation
upgrade unparsed alcohol tracker

baby-benz committed:
replace UpdateProducts class with new ParserApi class

baby-benz committed:
hotfix Strength type mismatch

baby-benz committed:
bugfix alter table migrations

baby-benz committed:
Merge branch 'sprint-6' into sprint-6-hotfix

baby-benz committed:
Merge pull request #16 from microservices-course-itmo/sprint-6-hotfix

Sprint 6 hotfix

baby-benz committed:
reformat code globally

kav128 committed:
add proxy to parser

kav128 committed:
resolve cherry-pick problems

kav128 committed:
proxy loading kostyl

kav128 committed:
split to multiple lines

kav128 committed:
remove Autowired Environment

kav128 committed:
Merge pull request #17 from microservices-course-itmo/proxy

Fix proxy

kanootoko committed:
Integration tests fixed

Use H2 instead of real postgres
Ensure that every test works as expected

baby-benz committed:
Merge pull request #23 from microservices-course-itmo/sprint-6-integration-tests

Integration tests fixed

kanootoko commented:
Sprint-12 is already merged to `dev`, so I'm closing this pull request
